pyrevolve/pyrevolve.py

```diff
@@ -115,107 +212,354 @@ def apply_reverse(self):
+
+        # remove storage list to avoid memory overflow
+        self.resetStorageList()
+        if diskstorage is True:
+            self.addDiskStorage(filedir=self.filedir, singlefile=self.singlefile)
+        else:
+            self.compression_params = compression_params
+            self.addNumpyStorage(compression_params)
 
-class MemoryRevolver(BaseRevolver):
+
+class MultiLevelRevolver(BaseRevolver):
     """
-        This class is an specialization of the Revolver class
-        that uses a NumpyStorage as its default storage method.
+    This class is an specialization of the Revolver class
+    that can use both NumpyStorage and Diskstorage as
+    storage methods. The HRevolve scheduler is used
+    to manage checkpointing with multiple storages.
+    It implements H-revolve algorithm as proposed in
+    the paper "H-Revolve: A Framework for Adjoint
+    Computationon Synchronous Hierarchical Platforms"
+    by Herrmann and Pallez [1]'
+
+    Ref:
+    [1] Herrmann, Pallez, "H-Revolve: A Framework for
+    Adjoint Computation on Synchronous Hierarchical
+    Platforms", ACM Transactions on Mathematical
+    Software  46(2), 2020.
     """
-    def __init__(self, checkpoint, fwd_operator, rev_operator,
-                 n_checkpoints=None, n_timesteps=None, timings=None,
-                 compression_params=None):
-        super().__init__(checkpoint,
-                         fwd_operator,
-                         rev_operator,
-                         n_checkpoints,
-                         n_timesteps,
-                         timings)
 
-        if compression_params is None:
-            compression_params = {'scheme': None}
+    def __init__(
+        self,
+        checkpoint,
+        fwd_operator,
+        rev_operator,
+        n_timesteps,
+        storage_list,
+        timings=None,
+        profiler=None,
+        uf=1,
+        ub=1,
+        up=1,
+    ):
+        """
+        Initializes a multi-level Revolver using HRevolve
+        scheduler. The user MUST provide a list of storage
+        objects. If an empty storage list is provided,
+        the scheduler attribute will not be created. In this
+        case, the user can add storages after the Revover
+        creation and reload the scheduler attribute.
 
-        self.compression_params = compression_params
-        self.createStorage()
+        @params:
+            checkpoint:         checkpoint object
+            fwd_operator:       forward operator
+            rev_operator:       backward operator
+            n_checkpoints:      number of checkpoints
+            n_timesteps:        number of timesteps
+            timings:            timings
+            profiler:           profiler
+            storage_list:       list of storage objects
+            uf:                 forward operation cost
+            uf:                 backward operation cost
+            uf:                 turn operation cost
```

Comment:
The above three should probably not be all `uf`

Reply:
My mistake.